test/hip_basic.cu

```diff
@@ -0,0 +1,170 @@
+// This file is part of Eigen, a lightweight C++ template library
+// for linear algebra.
+//
+// Copyright (C) 2015-2016 Gael Guennebaud <gael.guennebaud@inria.fr>
+//
+// This Source Code Form is subject to the terms of the Mozilla
+// Public License v. 2.0. If a copy of the MPL was not distributed
+// with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// workaround issue between gcc >= 4.7 and cuda 5.5
+#if (defined __GNUC__) && (__GNUC__>4 || __GNUC_MINOR__>=7)
+  #undef _GLIBCXX_ATOMIC_BUILTINS
+  #undef _GLIBCXX_USE_INT128
+#endif
+
+#define EIGEN_TEST_NO_LONGDOUBLE
+#define EIGEN_TEST_NO_COMPLEX
+#define EIGEN_TEST_FUNC hip_basic
+#define EIGEN_DEFAULT_DENSE_INDEX_TYPE int
+
+#include <hip/hip_runtime.h>
+
+#include "main.h"
+#include "hip_common.h"
+
+// Check that dense modules can be properly parsed by hipcc
+#include <Eigen/Dense>
+
+// struct Foo{
```

Comment:
is this struct commented out intentionally?

Reply:
yes...kinda

all *hip*.cu files are essentially copies of the corresponding *cuda*.cu files. So this commented out structure is something we inherited from the cuda_basic.cu file.  I can remove that if needed. The idea behind leaving it in was to avoid the number of diffs I had to visuallly parse when diffing the cuda and hip versions of the files to see things I might have missed.

Reply:
one thing I've been thinking is whether we keep a separate set of tests for `HIP`, or edit those `_cuda.cu` in-place. a large part of the PR is to introduce `HIP`-ified version of tests and i'm not sure who'd like to own/maintain it in the future.

Reply:
👍 if we can edit existing _cuda.cu in-place.  I agree it could help with future maintenance.

Reply:
I don't think we have choices here -- everything we contributed to the upstream Eigen, we'll need to maintain, including the unit tests. Besides, without unit tests, we won't be able to upstream our changes.  In-place edit will bring in more reviews from the CUDA maintainers, I don't think we like that.

Reply:
see discussion on same topic below...basically we will go with separate versions for hip unless the eigen-maintainers tell us otherwise